File: 06/walker.py
class MapWalker:
    DIRECTIONS = {
        '^': (-1, 0),  # up
        '>': (0, 1),   # right
        'v': (1, 0),   # down
        '<': (0, -1)   # left
    }
    
    TURN_RIGHT = {'<': '^', '^': '>', '>': 'v', 'v': '<'}

    def get_next_position(self, curr_row, curr_col, curr_dir):
        """Returns (next_row, next_col, next_dir, hit_wall, out_of_bounds)"""
        dr, dc = self.DIRECTIONS[curr_dir]
        next_row = curr_row + dr
        next_col = curr_col + dc
        
        # Check boundaries first
        if next_row < 0 or next_row >= self.height or next_col < 0 or next_col >= self.width:
            return next_row, next_col, curr_dir, False, True
            
        # Check for wall
        if self.map[next_row][next_col] == '#':
            return curr_row, curr_col, self.TURN_RIGHT[curr_dir], True, False
            
        return next_row, next_col, curr_dir, False, False

    def __init__(self, themap):
        self.map = themap
        self.height = len(themap)
        self.width = len(themap[0])
        self.start_pos = self._find_start()
        self.visited_states = set()
        
    def _find_start(self):
        for r in range(self.height):
            for c in range(self.width):
                if self.map[r][c] in '^>v<':
                    return (r, c)
        raise ValueError("No start position found")



    def walk(self):
        """Walk the map from the starting position.
        
        Returns:
            False if a loop is detected (same state visited twice)
            int: number of unique positions visited if the walk exits the map
            
        Raises:
            RuntimeError: if maximum steps are exceeded
        """
        curr_row, curr_col = self.start_pos
        curr_dir = self.map[curr_row][curr_col]
        self.visited_states.clear()
        steps = 0
        max_steps = self.height * self.width * 4
        
        while steps < max_steps:
            state = (curr_row, curr_col, curr_dir)
            
            if state in self.visited_states:
                return False
                
            self.visited_states.add(state)
            next_row, next_col, next_dir, hit_wall, out_of_bounds = self.get_next_position(curr_row, curr_col, curr_dir)
            
            if out_of_bounds:
                # Count unique positions from states - just take row,col from the (row,col,dir) tuples
                unique_positions = {(state[0], state[1]) for state in self.visited_states}
                return len(unique_positions)
                
            if hit_wall:
                curr_dir = next_dir
            else:
                curr_row, curr_col = next_row, next_col
                curr_dir = next_dir
            
            steps += 1
            
        raise RuntimeError(f"Exceeded maximum steps ({max_steps})")

File: 06/src/walker.rs
use std::collections::HashSet;

pub struct MapWalker {
    map: Vec<Vec<char>>,
    height: usize,
    width: usize,
    start_pos: (usize, usize),
    visited_states: HashSet<(usize, usize, char)>,
}

impl MapWalker {
    const DIRECTIONS: [(i32, i32); 4] = [(-1, 0), (0, 1), (1, 0), (0, -1)];

    pub fn new(map: &Vec<Vec<char>>) -> Self {
        let height = map.len();
        let width = map[0].len();
        let start_pos = Self::find_start(map);
        
        MapWalker {
            map: map.clone(),
            height,
            width,
            start_pos,
            visited_states: HashSet::new(),
        }
    }

    fn find_start(map: &Vec<Vec<char>>) -> (usize, usize) {
        for (r, row) in map.iter().enumerate() {
            for (c, &ch) in row.iter().enumerate() {
                if "^>v<".contains(ch) {
                    return (r, c);
                }
            }
        }
        panic!("No start position found");
    }

    fn get_next_position(&self, curr_row: usize, curr_col: usize, curr_dir: char) 
        -> (Option<(usize, usize)>, char, bool, bool) 
    {
        let (dr, dc) = match curr_dir {
            '^' => Self::DIRECTIONS[0],
            '>' => Self::DIRECTIONS[1],
            'v' => Self::DIRECTIONS[2],
            '<' => Self::DIRECTIONS[3],
            _ => panic!("Invalid direction"),
        };

        let next_row = curr_row as i32 + dr;
        let next_col = curr_col as i32 + dc;

        // Check boundaries
        if next_row < 0 || next_row >= self.height as i32 || 
           next_col < 0 || next_col >= self.width as i32 {
            return (None, curr_dir, false, true);
        }

        let next_row = next_row as usize;
        let next_col = next_col as usize;

        // Check for wall
        if self.map[next_row][next_col] == '#' {
            let next_dir = match curr_dir {
                '<' => '^',
                '^' => '>',
                '>' => 'v',
                'v' => '<',
                _ => panic!("Invalid direction"),
            };
            return (Some((curr_row, curr_col)), next_dir, true, false);
        }

        (Some((next_row, next_col)), curr_dir, false, false)
    }

    pub fn walk(&mut self) -> Result<Option<usize>, String> {
        let mut curr_pos = self.start_pos;
        let mut curr_dir = self.map[curr_pos.0][curr_pos.1];
        self.visited_states.clear();
        let max_steps = self.height * self.width * 4;
        let mut steps = 0;

        while steps < max_steps {
            let state = (curr_pos.0, curr_pos.1, curr_dir);
            
            if self.visited_states.contains(&state) {
                return Ok(None);
            }
            
            self.visited_states.insert(state);
            
            let (next_pos, next_dir, hit_wall, out_of_bounds) = 
                self.get_next_position(curr_pos.0, curr_pos.1, curr_dir);
            
            if out_of_bounds {
                let unique_positions: HashSet<(usize, usize)> = 
                    self.visited_states.iter().map(|&(r, c, _)| (r, c)).collect();
                return Ok(Some(unique_positions.len()));
            }
            
            curr_dir = next_dir;
            if !hit_wall {
                curr_pos = next_pos.unwrap();
            }
            
            steps += 1;
        }
        
        Err(format!("Exceeded maximum steps ({})", max_steps))
    }

    pub fn get_visited_positions(&self) -> HashSet<(usize, usize)> {
        self.visited_states.iter().map(|&(r, c, _)| (r, c)).collect()
    }
}
